File: app/src/main/java/hr/ml/moviesdiscover/view/MovieDetailsFragment.java
package hr.ml.moviesdiscover.view;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.List;

import hr.ml.moviesdiscover.R;
import hr.ml.moviesdiscover.adapter.MovieCastAdapter;
import hr.ml.moviesdiscover.model.Movie;
import hr.ml.moviesdiscover.rest.model.CastFromRetrofit;
import hr.ml.moviesdiscover.rest.model.MovieDetailsFromRetrofit;
import hr.ml.moviesdiscover.util.TmdbImageUrl;
import hr.ml.moviesdiscover.util.UiListToString;
import hr.ml.moviesdiscover.viewmodel.MovieDetailsViewModel;
import hr.ml.moviesdiscover.viewmodel.SharedDataViewModel;

public class MovieDetailsFragment extends Fragment {
    private static final String TAG = "MovieDetailsFragment";

    private ProgressBar progressBarActors;
    private ProgressBar progressBarBoxOffice;
    private ImageView movieImage;
    private TextView movieTitle;
    private TextView movieYear;
    private TextView movieGenres;
    private TextView movieRating;
    private TextView movieDescription;
    private TextView movieRevenueLabel;
    private TextView movieRevenue;
    private TextView movieBudgetLabel;
    private TextView movieBudget;
    private RecyclerView castRecyclerView;

    private SharedDataViewModel sharedViewModel;
    private MovieDetailsViewModel viewModel;

    public MovieDetailsFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_movie_details, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // fragment changes appbar
        setHasOptionsMenu(true);

        sharedViewModel = new ViewModelProvider(requireActivity()).get(SharedDataViewModel.class);

        viewModel = new ViewModelProvider(this).get(MovieDetailsViewModel.class);

        progressBarActors = view.findViewById(R.id.progress_actors);
        progressBarBoxOffice = view.findViewById(R.id.progress_box_office);
        movieImage = view.findViewById(R.id.movie_image);
        movieTitle = view.findViewById(R.id.movie_title);
        movieYear = view.findViewById(R.id.movie_year);
        movieGenres = view.findViewById(R.id.movie_genres);
        movieRating = view.findViewById(R.id.movie_rating);
        movieDescription = view.findViewById(R.id.movie_description);
        castRecyclerView = view.findViewById(R.id.recycler_view_cast);
        movieRevenueLabel = view.findViewById(R.id.movie_revenue_label);
        movieRevenue = view.findViewById(R.id.movie_revenue);
        movieBudgetLabel = view.findViewById(R.id.movie_budget_label);
        movieBudget = view.findViewById(R.id.movie_budget);

        MovieCastAdapter castAdapter = new MovieCastAdapter();
        castRecyclerView.setAdapter(castAdapter);

        sharedViewModel.selectedMovie.observe(requireActivity(), movie -> {
            updateUiWith(movie);
        });

        viewModel.movieDetails.observe(getViewLifecycleOwner(), movieDetailsFromRetrofit -> {
            progressBarBoxOffice.setVisibility(View.GONE);

            if(movieDetailsFromRetrofit != null) {
                updateUiWith(movieDetailsFromRetrofit);
            } else Log.d(TAG, "request_failed");
        });

        viewModel.movieCast.observe(getViewLifecycleOwner(), new Observer<List<CastFromRetrofit>>() {
            @Override
            public void onChanged(List<CastFromRetrofit> castFromRetrofits) {
                progressBarActors.setVisibility(View.GONE);

                if(castFromRetrofits != null) castAdapter.submitList(castFromRetrofits);
                else Log.d(TAG, "request_failed");
            }
        });
    }

    private void updateUiWith(Movie movie) {
        viewModel.requestMoviesDetails(movie.getId());
        viewModel.requestMovieCast(movie.getId());
        movieTitle.setText(String.valueOf(movie.getTitle()));
        movieYear.setText("(" + movie.getReleaseDate() + ")");
        movieRating.setText(movie.getVoteAverage() + "/10\nVotes: " + movie.getVoteCount());
        movieDescription.setText(movie.getOverview());
        String imageUrl = TmdbImageUrl
                .generatePosterUrl(movie.getPosterPath(), TmdbImageUrl.PosterWidth.w342);
        Picasso.get()
                .load(imageUrl)
                .placeholder(R.drawable.ic_outline_image)
                .into(movieImage);
    }

    private void updateUiWith(MovieDetailsFromRetrofit movieDetailsFromRetrofit) {
        movieGenres.setText(UiListToString.genresToString(movieDetailsFromRetrofit.getGenres()));

        if(movieDetailsFromRetrofit.getBudget() > 0) {
            movieBudgetLabel.setVisibility(View.VISIBLE);
            movieBudget.setText("$" + movieDetailsFromRetrofit.getBudget());
        }

        if(movieDetailsFromRetrofit.getRevenue() > 0) {
            movieRevenueLabel.setVisibility(View.VISIBLE);
            movieRevenue.setText("$" + movieDetailsFromRetrofit.getRevenue());
        }
    }


    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_movie_details, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.share_movie:
                shareMovie();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void shareMovie() {
        Movie movieToShare = sharedViewModel.selectedMovie.getValue();

        Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
        sharingIntent.setType("text/plain");
        String shareBody = "Check out movie: " + movieToShare.getTitle()
                + " (" + movieToShare.getReleaseDate() + ")!";
        sharingIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, "Discover movie");
        sharingIntent.putExtra(android.content.Intent.EXTRA_TEXT, shareBody);
        startActivity(Intent.createChooser(sharingIntent, "Share via"));
    }
}

File: app/src/main/kotlin/hr/ml/moviesdiscover/view/MovieDetailsFragment.kt
package hr.ml.moviesdiscover.view

import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import hr.ml.moviesdiscover.R
import hr.ml.moviesdiscover.adapter.MovieCastAdapter
import hr.ml.moviesdiscover.model.Movie
import hr.ml.moviesdiscover.rest.model.MovieDetailsFromRetrofit
import hr.ml.moviesdiscover.util.TmdbImageUrl
import hr.ml.moviesdiscover.util.TmdbImageUrl.Companion.generatePosterUrl
import hr.ml.moviesdiscover.util.UiListToString
import hr.ml.moviesdiscover.viewmodel.MovieDetailsViewModel
import hr.ml.moviesdiscover.viewmodel.SharedDataViewModel

class MovieDetailsFragment : Fragment() {

    private lateinit var progressBarActors: ProgressBar
    private lateinit var progressBarBoxOffice: ProgressBar
    private lateinit var movieImage: ImageView
    private lateinit var movieTitle: TextView
    private lateinit var movieYear: TextView
    private lateinit var movieGenres: TextView
    private lateinit var movieRating: TextView
    private lateinit var movieDescription: TextView
    private lateinit var movieRevenueLabel: TextView
    private lateinit var movieRevenue: TextView
    private lateinit var movieBudgetLabel: TextView
    private lateinit var movieBudget: TextView
    private lateinit var castRecyclerView: RecyclerView

    private lateinit var sharedViewModel: SharedDataViewModel
    private lateinit var viewModel: MovieDetailsViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_movie_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setHasOptionsMenu(true)

        sharedViewModel = ViewModelProvider(requireActivity()).get(SharedDataViewModel::class.java)
        viewModel = ViewModelProvider(this).get(MovieDetailsViewModel::class.java)

        progressBarActors = view.findViewById(R.id.progress_actors)
        progressBarBoxOffice = view.findViewById(R.id.progress_box_office)
        movieImage = view.findViewById(R.id.movie_image)
        movieTitle = view.findViewById(R.id.movie_title)
        movieYear = view.findViewById(R.id.movie_year)
        movieGenres = view.findViewById(R.id.movie_genres)
        movieRating = view.findViewById(R.id.movie_rating)
        movieDescription = view.findViewById(R.id.movie_description)
        castRecyclerView = view.findViewById(R.id.recycler_view_cast)
        movieRevenueLabel = view.findViewById(R.id.movie_revenue_label)
        movieRevenue = view.findViewById(R.id.movie_revenue)
        movieBudgetLabel = view.findViewById(R.id.movie_budget_label)
        movieBudget = view.findViewById(R.id.movie_budget)

        val adapter = MovieCastAdapter()
        castRecyclerView.adapter = adapter

        sharedViewModel.selectedMovie.observe(requireActivity(), { movie ->
            updateUIWith(movie)
        })

        viewModel.movieDetails.observe(viewLifecycleOwner, {
            progressBarBoxOffice.visibility = View.GONE
            if (it != null) updateUiWithDetails(it)
        })

        viewModel.movieCast.observe(viewLifecycleOwner, {
            progressBarActors.visibility = View.GONE
            adapter.submitList(it)
        })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_movie_details, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.share_movie -> {
                shareMovie()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun updateUIWith(movie: Movie) {
        viewModel.requestMovieCast(movie.id)
        viewModel.requestMovieDetails(movie.id)
        movieTitle.text = movie.title
        movieYear.text = "(${movie.releaseDate})"
        movieRating.text = "${movie.voteAverage}/10\nVotes: ${movie.voteCount}"
        movieDescription.text = movie.overview

        val imageUrl = generatePosterUrl(movie.posterPath, TmdbImageUrl.PosterWidth.W342)
        Picasso.get()
                .load(imageUrl)
                .placeholder(R.drawable.ic_outline_image)
                .into(movieImage)
    }

    private fun updateUiWithDetails(movieDetailsFromRetrofit: MovieDetailsFromRetrofit) {
        movieGenres.text = UiListToString.genresToString(movieDetailsFromRetrofit.genres.orEmpty())

        if (movieDetailsFromRetrofit.budget > 0) {
            movieBudgetLabel.visibility = View.VISIBLE
            movieBudget.text = "\$${movieDetailsFromRetrofit.budget}"
        }

        if (movieDetailsFromRetrofit.revenue > 0) {
            movieRevenueLabel.visibility = View.VISIBLE
            movieRevenue.text = "\$${movieDetailsFromRetrofit.revenue}"
        }
    }

    private fun shareMovie() {
        val movieToShare = sharedViewModel.selectedMovie.value

        val sharingIntent = Intent(Intent.ACTION_SEND)
        sharingIntent.type = "text/plain"
        val shareBody = "Check out movie: ${movieToShare?.title} (${movieToShare?.releaseDate})!"
        sharingIntent.putExtra(Intent.EXTRA_SUBJECT, "Discover movie")
        sharingIntent.putExtra(Intent.EXTRA_TEXT, shareBody)
        startActivity(Intent.createChooser(sharingIntent, "Share via"))
    }
}